[SPARK-5337][Standalone] respect spark.task.cpus when scheduling Applications

CodingCat opened the pull request:
https://issues.apache.org/jira/browse/SPARK-5337

Currently, we didn't consider spark.task.cpus when scheduling the applications in Master, so that we may fall into one of the following cases

the executor gets N cores but we need M cores to run a single task, where N < M

the executor gets N cores, we need M cores to run a single task, where N % M != 0 && N > M; so that we waste some cores in the executor

Patch for YARN is in submitted by @WangTaoTheTonic : #4123

CodingCat committed:
respect spark.task.cpus when scheduling Applications

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42037 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42037/consoleFull) for   PR 8610 at commit [`d3289fc`](https://github.com/apache/spark/commit/d3289fc331bd2e6bb19453f3489ff3ca4d865c90).

SparkQA commented:
[Test build #42037 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42037/console) for   PR 8610 at commit [`d3289fc`](https://github.com/apache/spark/commit/d3289fc331bd2e6bb19453f3489ff3ca4d865c90).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42037/
Test FAILed.

CodingCat commented:
retest this please

CodingCat commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

CodingCat commented:
....first time to know I have the permission to `retest` stuffs..............

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42054 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42054/consoleFull) for   PR 8610 at commit [`d3289fc`](https://github.com/apache/spark/commit/d3289fc331bd2e6bb19453f3489ff3ca4d865c90).

SparkQA commented:
[Test build #42055 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42055/consoleFull) for   PR 8610 at commit [`d3289fc`](https://github.com/apache/spark/commit/d3289fc331bd2e6bb19453f3489ff3ca4d865c90).

SparkQA commented:
[Test build #42055 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42055/console) for   PR 8610 at commit [`d3289fc`](https://github.com/apache/spark/commit/d3289fc331bd2e6bb19453f3489ff3ca4d865c90).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42055/
Test FAILed.

SparkQA commented:
[Test build #42054 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42054/console) for   PR 8610 at commit [`d3289fc`](https://github.com/apache/spark/commit/d3289fc331bd2e6bb19453f3489ff3ca4d865c90).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42054/
Test FAILed.

CodingCat committed:
fix compilation error

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42059 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42059/consoleFull) for   PR 8610 at commit [`9c16dc7`](https://github.com/apache/spark/commit/9c16dc78c32d86444de9a88f01ba897edd9af87a).

SparkQA commented:
[Test build #42059 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42059/console) for   PR 8610 at commit [`9c16dc7`](https://github.com/apache/spark/commit/9c16dc78c32d86444de9a88f01ba897edd9af87a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42059/
Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

CodingCat committed:
fix Mesos issue by assigning CPUS_PER_TASK a valid value

SparkQA commented:
[Test build #42060 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42060/consoleFull) for   PR 8610 at commit [`0871c37`](https://github.com/apache/spark/commit/0871c37e53739d529181d754c2a0dfde53ade9b2).

SparkQA commented:
[Test build #42060 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42060/console) for   PR 8610 at commit [`0871c37`](https://github.com/apache/spark/commit/0871c37e53739d529181d754c2a0dfde53ade9b2).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42060/
Test PASSed.

CodingCat commented:
Hi, @andrewor14 , I just added some test cases here, would you mind taking the review?

CodingCat committed:
add test cases

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42094 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42094/consoleFull) for   PR 8610 at commit [`44c6a03`](https://github.com/apache/spark/commit/44c6a03a4295db38e93cff150bc3ea84e0b7fa66).

SparkQA commented:
[Test build #42094 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42094/console) for   PR 8610 at commit [`44c6a03`](https://github.com/apache/spark/commit/44c6a03a4295db38e93cff150bc3ea84e0b7fa66).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42094/
Test PASSed.

holdenk commented on core/src/main/scala/org/apache/spark/scheduler/cluster/SparkDeploySchedulerBackend.scala:
```diff
@@ -83,8 +83,16 @@ private[spark] class SparkDeploySchedulerBackend(
+      throw new SparkException(s"invalid configuration of " +
+        s"spark.executor.cores(${coresPerExecutor.get}) and " +
+        s"spark.task.cpus(${scheduler.CPUS_PER_TASK}}), spark.executor.cores has to be no less " +
+        "than and folds of spark.task.cpus")
```
I don't think this error message is super clear, what is meant by "no less than and folds of spark.task.cups"?

CodingCat replied:
@holdenk how about now?

holdenk replied:
clearer, thanks :)

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

CodingCat committed:
clearer comments

SparkQA commented:
[Test build #42211 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42211/consoleFull) for   PR 8610 at commit [`169141a`](https://github.com/apache/spark/commit/169141a36c5770f7cca1591e89ac61d2627461ab).

SparkQA commented:
[Test build #42211 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42211/console) for   PR 8610 at commit [`169141a`](https://github.com/apache/spark/commit/169141a36c5770f7cca1591e89ac61d2627461ab).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42211/
Test FAILed.

CodingCat commented:
Jenkins, retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42217 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42217/consoleFull) for   PR 8610 at commit [`169141a`](https://github.com/apache/spark/commit/169141a36c5770f7cca1591e89ac61d2627461ab).

SparkQA commented:
[Test build #42217 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42217/console) for   PR 8610 at commit [`169141a`](https://github.com/apache/spark/commit/169141a36c5770f7cca1591e89ac61d2627461ab).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42217/
Test FAILed.

CodingCat commented:
Jenkins, retest it please

CodingCat commented:
Jenkins, retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42229 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42229/consoleFull) for   PR 8610 at commit [`169141a`](https://github.com/apache/spark/commit/169141a36c5770f7cca1591e89ac61d2627461ab).

SparkQA commented:
[Test build #42229 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42229/console) for   PR 8610 at commit [`169141a`](https://github.com/apache/spark/commit/169141a36c5770f7cca1591e89ac61d2627461ab).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42229/
Test PASSed.

CodingCat commented:
@andrewor14  would you have some chance to review this?

dragos commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -639,10 +640,11 @@ private[deploy] class Master(
     // in the queue, then the second app, etc.
     for (app <- waitingApps if app.coresLeft > 0) {
       val coresPerExecutor: Option[Int] = app.desc.coresPerExecutor
+      val coreNumPerTask = app.desc.coresPerTask
```
nitpick: `coreNumPerTask` sounds weird.  `coresPerTaks` is what you've used everywhere else, and the name above is `coresPerExecutor`.

SparkQA commented:
**[Test build #46536 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46536/consoleFull)** for PR 8610 at commit [`f27f5d8`](https://github.com/apache/spark/commit/f27f5d88e30b31193ef6b45323eb1a60c6e64b49).

CodingCat committed:
rename var

SparkQA commented:
**[Test build #46536 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46536/consoleFull)** for PR 8610 at commit [`f27f5d8`](https://github.com/apache/spark/commit/f27f5d88e30b31193ef6b45323eb1a60c6e64b49).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46536/
Test FAILed.

CodingCat commented:
Jenkins, retest this please

CodingCat commented:
Jenkins, retest this please

CodingCat commented:
Hi, @dragos , thanks for the comments, just sync the patch with master and addressed your comments

@andrewor14 any plan to fix this in the coming 1.6?

andrewor14 commented:
The 1.6 preview is already cut. This will have to come later. Until then, it would be good for @tnachen and @dragos to have a look.

CodingCat commented:
OK, thx

One more question, am I supposed to have the permission to trigger Jenkins to retest the patch, I thought  I was once able to do that...or it is due to the weird status of the bot?

CodingCat commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #46549 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46549/consoleFull)** for PR 8610 at commit [`f27f5d8`](https://github.com/apache/spark/commit/f27f5d88e30b31193ef6b45323eb1a60c6e64b49).

CodingCat commented:
nvm, the bot fixed the relationship with me .....

tnachen commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -639,10 +640,11 @@ private[deploy] class Master(
       val usableWorkers = workers.toArray.filter(_.state == WorkerState.ALIVE)
         .filter(worker => worker.memoryFree >= app.desc.memoryPerExecutorMB &&
-          worker.coresFree >= coresPerExecutor.getOrElse(1))
+          worker.coresFree >= math.max(coresPerExecutor.getOrElse(1), coresPerTask))
```
IMO it's not clear what the relationship between coresPerTask and coresPerExecutor. Looking at the scheduling requirement we just need the maximum of the set to be able to schedule, and I thought cores per task is additional cpu resources on top of the executor. 
Can we perhaps comment where this is introduced what these two are? Or point to documentation?

SparkQA commented:
**[Test build #46549 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46549/consoleFull)** for PR 8610 at commit [`f27f5d8`](https://github.com/apache/spark/commit/f27f5d88e30b31193ef6b45323eb1a60c6e64b49).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46549/
Test PASSed.

dragos commented:
I'll have a look at it tomorrow, thanks for pinging me.

CodingCat commented:
@dragos @tnachen  thanks for reviewing, I will handle your comments tomorrow

CodingCat committed:
add comments about Spark-5337

SparkQA commented:
**[Test build #46768 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46768/consoleFull)** for PR 8610 at commit [`8232a80`](https://github.com/apache/spark/commit/8232a808e398f3644304822d1824aa0b923090dc).

SparkQA commented:
**[Test build #46768 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46768/consoleFull)** for PR 8610 at commit [`8232a80`](https://github.com/apache/spark/commit/8232a808e398f3644304822d1824aa0b923090dc).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46768/
Test FAILed.

CodingCat commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #46774 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46774/consoleFull)** for PR 8610 at commit [`8232a80`](https://github.com/apache/spark/commit/8232a808e398f3644304822d1824aa0b923090dc).

SparkQA commented:
**[Test build #46774 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46774/consoleFull)** for PR 8610 at commit [`8232a80`](https://github.com/apache/spark/commit/8232a808e398f3644304822d1824aa0b923090dc).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46774/
Test FAILed.

CodingCat commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #46785 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46785/consoleFull)** for PR 8610 at commit [`8232a80`](https://github.com/apache/spark/commit/8232a808e398f3644304822d1824aa0b923090dc).

CodingCat commented:
weird, something magical happened after Nov 23th

CodingCat commented:
Leave a mark here, there is something wrong with a particular test case causing the above failures, more details:   https://issues.apache.org/jira/browse/SPARK-12021?jql=project%20%3D%20SPARK

SparkQA commented:
**[Test build #46785 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46785/consoleFull)** for PR 8610 at commit [`8232a80`](https://github.com/apache/spark/commit/8232a808e398f3644304822d1824aa0b923090dc).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46785/
Test FAILed.

CodingCat commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #46829 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46829/consoleFull)** for PR 8610 at commit [`8232a80`](https://github.com/apache/spark/commit/8232a808e398f3644304822d1824aa0b923090dc).

SparkQA commented:
**[Test build #46829 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46829/consoleFull)** for PR 8610 at commit [`8232a80`](https://github.com/apache/spark/commit/8232a808e398f3644304822d1824aa0b923090dc).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46829/
Test PASSed.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -571,7 +571,8 @@ private[deploy] class Master(
     val coresPerExecutor = app.desc.coresPerExecutor
-    val minCoresPerExecutor = coresPerExecutor.getOrElse(1)
+    val coresPerTask = app.desc.coresPerTask
+    val minCoresPerExecutor = math.max(coresPerExecutor.getOrElse(1), coresPerTask)
```
I don't think we should do this. If the user explicitly specified `spark.executor.cores` they expect their executors to be scheduled with that many cores, no more no less.

andrewor14 commented:
@CodingCat thanks for working on this. Sorry that we haven't had the bandwidth to review this until now.

For this patch, I think it's better if we just throw an exception if `spark.task.cpus > spark.executor.cores`. We shouldn't start launching executors with more cores silently just because `spark.task.cpus` is big. That is not what I would expect as a user.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/ApplicationDescription.scala:
```diff
@@ -26,10 +26,10 @@ private[spark] case class ApplicationDescription(
     command: Command,
     appUiUrl: String,
     eventLogDir: Option[URI] = None,
-    // short name of compression codec used when writing event logs, if any (e.g. lzf)
```
don't remove this...

andrewor14 commented:
By the way, I think it's totally fine to just do this for standalone mode first, since Mesos doesn't yet read `spark.executor.cores`. We can fix it for Mesos later.

CodingCat committed:
stylistic fix

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50713/
Test FAILed.

SparkQA commented:
**[Test build #50715 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50715/consoleFull)** for PR 8610 at commit [`db57a2e`](https://github.com/apache/spark/commit/db57a2e358e3e158338d531a4600be4a8cbc6a59).

CodingCat committed:
do not start executors with changed perCore Configuration

SparkQA commented:
**[Test build #50715 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50715/consoleFull)** for PR 8610 at commit [`db57a2e`](https://github.com/apache/spark/commit/db57a2e358e3e158338d531a4600be4a8cbc6a59).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50715/
Test FAILed.

SparkQA commented:
**[Test build #50719 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50719/consoleFull)** for PR 8610 at commit [`ae3e5e4`](https://github.com/apache/spark/commit/ae3e5e490aee08afde96875d91d16af60130c187).

CodingCat commented:
@andrewor14 thanks for reviewing this, I'm currently in traveling, I will move https://github.com/apache/spark/pull/4123 and the check here to SparkSubmit later 

I think we only need to check whether CoresPerExecutor is no less than CoresPerTask `only when the user has an explicit configuration on coresPerExecutor`

The reason is that when the user does not explicitly specify coresPerExecutor (default is 1), there are still `many chances that executors has more cores than coresPerTask`...if we simply throw exceptions to the user in this case, the user has to set an explicit value of coresPerExecutor `whenever they set a value on coresPerTask` which is not an ideal situtation in terms of user-friendly , your thoughts?

SparkQA commented:
**[Test build #50719 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50719/consoleFull)** for PR 8610 at commit [`ae3e5e4`](https://github.com/apache/spark/commit/ae3e5e490aee08afde96875d91d16af60130c187).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50719/
Test FAILed.

CodingCat committed:
compilation fix

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50724/
Test FAILed.

CodingCat commented:
Jenkins, retest it please

CodingCat commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #50743 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50743/consoleFull)** for PR 8610 at commit [`e50e705`](https://github.com/apache/spark/commit/e50e705014084de3c3a5e9738a2bc6b203616a79).

SparkQA commented:
**[Test build #50743 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50743/consoleFull)** for PR 8610 at commit [`e50e705`](https://github.com/apache/spark/commit/e50e705014084de3c3a5e9738a2bc6b203616a79).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50743/
Test FAILed.

CodingCat committed:
distinguish for oneExecutorPerWorker and multiExecutorsPerWorker

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50816/
Test FAILed.

SparkQA commented:
**[Test build #50817 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50817/consoleFull)** for PR 8610 at commit [`8286e97`](https://github.com/apache/spark/commit/8286e972fa43135c2d7dad4f103b77d55b72e163).

SparkQA commented:
**[Test build #50817 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50817/consoleFull)** for PR 8610 at commit [`8286e97`](https://github.com/apache/spark/commit/8286e972fa43135c2d7dad4f103b77d55b72e163).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50817/
Test FAILed.

CodingCat committed:
recover initialExecutorLimit

SparkQA commented:
**[Test build #50847 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50847/consoleFull)** for PR 8610 at commit [`d644654`](https://github.com/apache/spark/commit/d644654d3af68c2c9f53978ed81f19da737460c5).

CodingCat commented:
Hi, @andrewor14, when I had a second think about your suggestion that we shall put parameter checking in SparkSubmit, I found that it may not be the right way to do this

Because the users are always able to set the parameters by code, instead of through spark-submit. Reading the parameters set in a programmable way happens after reading the parameters set through spark-submit. In this case, we have to check whether the parameters are making sense right before we use them to create the application descriptor, e.g. ApplicationDescription in standalone mode. Otherwise, we may miss the parameters which are set after spark-submit parameter reading...

SparkQA commented:
**[Test build #50847 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50847/consoleFull)** for PR 8610 at commit [`d644654`](https://github.com/apache/spark/commit/d644654d3af68c2c9f53978ed81f19da737460c5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50847/
Test FAILed.

CodingCat commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #50853 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50853/consoleFull)** for PR 8610 at commit [`d644654`](https://github.com/apache/spark/commit/d644654d3af68c2c9f53978ed81f19da737460c5).

SparkQA commented:
**[Test build #50853 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50853/consoleFull)** for PR 8610 at commit [`d644654`](https://github.com/apache/spark/commit/d644654d3af68c2c9f53978ed81f19da737460c5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50853/
Test PASSed.

rxin commented:
Thanks for the pull request. I'm going through a list of pull requests to cut them down since the sheer number is breaking some of the tooling we have. Due to lack of activity on this pull request, I'm going to push a commit to close it. Feel free to reopen it or create a new one. We can also continue the discussion on the JIRA ticket.
